rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+
+### Audit log
+
+The following actions will be tracked with audit events:
+
+* User initiates headless login
+* User approves/denies headless login request
+
+### Server changes
+
+Headless authentication has a unique API flow compared to other login methods.
+
+#### `POST /webapi/login/headless`
+
+This endpoint is used to initiate headless login. Like other login endpoints, this endpoint is not authenticated and can be called by anyone with access to the Teleport Proxy address.
+
+```go
+type HeadlessAuthRequest struct {
+    SSHLogin
+    // User is a teleport username
+    User string `json:"user"`
+    // RequestID is a uuid for the request
+    RequestID string
+    // Command is the client command being requested
+    Command string
+}
+
+// SSHLogin contains common SSH login parameters.
+type SSHLogin struct {
+    // ProxyAddr is the target proxy address
+    ProxyAddr string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    ...
+}
+
+// SSHLoginResponse is a user login response
+type SSHLoginResponse struct {
+    // Username contains the username for the login certificates
+    Username string `json:"username"`
+    // Cert is a PEM encoded SSH certificate signed by SSH certificate authority
+    Cert []byte `json:"cert"`
+    // TLSCertPEM is a PEM encoded TLS certificate signed by TLS certificate authority
+    TLSCert []byte `json:"tls_cert"`
+    // HostSigners is a list of signing host public keys trusted by proxy
+    HostSigners []TrustedCerts `json:"host_signers"`
+}
+```
+
+#### `rpc CreateHeadlessAuthRequest`
+
+This endpoint is used by the Teleport Proxy to create a new headless auth request.
+
+```proto
+service AuthService {
+  rpc CreateHeadlessAuthRequest(CreateHeadlessAuthRequestRequest) returns (HeadlessAuthRequest);
+}
+
+message CreateHeadlessAuthRequestRequest {
+  // RequestID is the headless login request uuid
+  string request_id = 1;
+}
+
+message HeadlessAuthRequest {
+  // ID is the headless login request uuid
+  string id = 1;
+  // User is a teleport user name
+  string user = 2;
+   // Command is a `tsh` command in plain text
+  string command = 3;
+  // State is the headless login request state
+  State state = 4;
+  // PublicKey is a public key to sign in case of successful auth.
+  bytes public_key = 5;
+  // MFADevice is the mfa device used to approve the request in case of successful auth.
+  MFADevice mfa_device = 6;
+  // Compatibility specifies OpenSSH compatibility flags.
+  string compatibility = 7;
+  // Expires is a global expiry time header can be set on any resource in the system.
+  google.protobuf.Timestamp expires = 8;
+  // RouteToCluster is the name of Teleport cluster to issue credentials for.
+  string route_to_cluster = 9;
+}
+
+// State is a headless login request state
+enum State {
+  NONE = 0;
+  PENDING = 1;
+  DENIED = 2;
+  APPOVED = 3;
+}
```

Comment:
```suggestion
enum HeadlessAuthenticationState {
  HEADLESS_AUTHENTICATION_STATE_UNSPECIFIED = 0;
  HEADLESS_AUTHENTICATION_STATE_PENDING = 1;
  HEADLESS_AUTHENTICATION_STATE_DENIED = 2;
  HEADLESS_AUTHENTICATION_STATE_APPOVED = 3;
}
```

Enums names go to the global variable namespace in some generated languages, so it's important to avoid being too-generic with their names.

https://protobuf.dev/programming-guides/style/#enums